### DISCUSSION TITLE
RIVER LEAKAGE and FLOW LOWER FACE not found in the budget file

### DISCUSSION THREAD

**Discussion by QingHe-UT** [modflowpy/flopy#1707] [Q&A]
Hi all,

Following a previous discussion (https://github.com/modflowpy/flopy/discussions/1704), I added a river object to the model. The river model is built as:
`#adding river`
`riv_row = [0,1,1,2,3,3,4,5,6,6,7,8,9]`
`riv_col = [4,4,5,5,5,6,6,6,6,7,7,7,7]`

`riv_len = delr    #river length`
`riv_wdt = delc    #river width`
`riv_bdp = 10.     #river bed depth`
`riv_bk  = 5e-5    #hydraulic conductivity of river bed`
`riv_rbot = 4     #river bottom`

`riv_cond = riv_bk*riv_len*riv_wdt/riv_bdp   #conductance of the riverbed material`

`#stress period 1`
`riv_sp1 = []`
`stage_sp1 = 10    #river head at stress period 1`
`for inum in range(len(riv_row)):`
    `riv_sp1.append([0, riv_row[inum], riv_col[inum], stage_sp1, riv_cond, riv_rbot])`
`print("Adding ", len(riv_sp1), "Stages for stress period 1.")`

`#stress period 2`
`riv_sp2 = []`
`stage_sp2 = 12    #river head at stress period 1`
`for inum in range(len(riv_row)):`
    `riv_sp2.append([0, riv_row[inum], riv_col[inum], stage_sp2, riv_cond, riv_rbot])`
`print("Adding ", len(riv_sp2), "Stages for stress period 2.")`

`stress_period_data = {0: riv_sp1, 1: riv_sp2}`

`#create river obj`
`riv = fp.modflow.ModflowRiv(sim,`
                            `ipackcb = 1,`
                            `stress_period_data = stress_period_data)`

The model is run successfully but I tried to extract the river leakage and flow lower face from the budget file, they cannot be found by the code I use (see below). I wonder is this because the river model was not added to the simulation, or some other reason (wrong post-processing code for example)?

**The processing code I use is:**
`cbbobj = fp.utils.CellBudgetFile("./toy_tvar_mf05/modflowtest.cbc")`
`times = cbbobj.get_times()`
`leak = cbbobj.get_data(text = "RIVER LEAKAGE", totim = times[-1])[0]`
`face_lower = cbbobj.get_data(text = "flow lower face", totim = times[-1])[0]`

**The error information is:**
`Cell In[328], line 3`
`1 cbbobj = fp.utils.CellBudgetFile("./toy_tvar_mf05/modflowtest.cbc")`
      `2 times = cbbobj.get_times()`
`----> 3 leak = cbbobj.get_data(text = "RIVER LEAKAGE", totim = times[-1])[0]`
      `4 #cbbobj.list_records()`
      `5 #face_lower = cbbobj.get_data(text = "flow lower face", totim = times[-1])[0]`

`File /Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/flopy/utils/binaryfile.py:1271, in` `CellBudgetFile.get_data(self, idx, kstpkper, totim, text, paknam, full3D)`
   `1269 text16 = None`
   `1270 if text is not None:`
`-> 1271     text16 = self._find_text(text)`
   `1272 paknam16 = None`
   `1273 if paknam is not None:`

`File /Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/flopy/utils/binaryfile.py:1020, in` `CellBudgetFile._find_text(self, text)`
   `1018     if text16 is None:`
   `1019         errmsg = "The specified text string is not in the budget file."`
`-> 1020         raise Exception(errmsg)`
   `1021 return text16`

`Exception: The specified text string is not in the budget file.`

I'm very new to MODFLOW so do forgive me for my massive postings recently (and many of them are very basic things...) I'm very much appreciated for your kind help!

**Comment by jlarsen-usgs** [comment]
@QingHe-UT 

When you construct your `ModflowGwf` model object have you set the `save_flows` parameter to `True`? That might explain why you are not seeing flows in your binary budget file. Example:

```python
sim = flopy.mf6.MFSimulation()
gwf = flopy.mf6.ModflowGwf(sim, modelname="my model", save_flows=True)

```
you can also get a list of the unique record names in your budget file by calling `get_unique_record_names`. Example:

```python
cbc = gwf.output.budget()
rec_names = cbc.get_unique_record_names()
```

**Reply by QingHe-UT** [reply]
Yes, the `save_flow` parameter has been set to `True` in the model. 

But I realize my simulation was conducted using MF2005, and there are some differences between MF2005 and MF6. 
In MF2005, the budget file contains record names as:
`[b'   CONSTANT HEAD', b'FLOW RIGHT FACE ', b'FLOW FRONT FACE ', b'         STORAGE']`
But in MF6, the record names are:
`[b'    STO-SS', b'    FLOW-JA-FACE', b'  DATA-SPDIS', b'  RIV', b'  GHB', b'   RCHA']`
So it seems the visualization code should be changed when post-processing results from MF6.

Regarding this point, is there any tutorial about how to visualize MF6 output? I only see some very simple descriptions in [Accessing MODFLOW 6 Output](https://flopy.readthedocs.io/en/3.3.6/_notebooks/tutorial01_mf6_output.html). But as a beginner in this field , sometimes I cannot fully understand how to use these functions exactly. Is it possible to give us more examples?

**Reply by QingHe-UT** [reply]
For example, I know the river leakage may be stored in `RIV` of the budget file, and it prints like this:
`leak = cbc.get_data(text = "RIV", totim = times[-1])[0]`
`print(leak)`
`[( 5,  1, 0.26582129) (15,  2, 0.26582065) (16,  3, 0.3200982 )`
` (26,  4, 0.32008235) (36,  5, 0.32008242) (37,  6, 0.38010479)`
` (47,  7, 0.38009134) (57,  8, 0.38009142) (67,  9, 0.38010319)`
` (68, 10, 0.4       ) (78, 11, 0.4       ) (88, 12, 0.4       )`
` (98, 13, 0.4       )]`

But how can I visualize the number above? I'm really struggling on this :(

**Reply by jlarsen-usgs** [reply]
For Modflow 2005 based models, the parameter `ipakcb` needs to be set for each package to the same unit number as your binary cell budget file. FloPy's default binary cell budget file unit number is 53.  Example:

```python
riv = fp.modflow.ModflowRiv(sim,
ipackcb=53,
stress_period_data=stress_period_data)
``` 

As far as visualization is concerned, you can get budget output as a 3d array and then plot it using FloPy's `PlotMapView()`:

```python
import matplotlib.pyplot as plt

leak = cbc.get_data(text="RIV", totim=times[-1], full3D=True)

pmv = flopy.plot.PlotMapView(model=gwf, layer=0)
pc = pmv.plot_array(leak, masked_values=[0,])
plt.colorbar(pc)
plt.show()
```

**Reply by QingHe-UT** [reply]
Thank you for your reply! I tried the code above, but got an error as:
`Exception Traceback (most recent call last)`
`Cell In[429], line 9`
     ` 6 print(leak)`
      `8 pmv = fp.plot.PlotMapView(model=gwf, layer=0)`
`----> 9 pc = pmv.plot_array(leak, masked_values=[0,])`
     `10 #plt.colorbar(pc)`
     `11 #plt.show()`
     `12   `
     `13 #pmv = fp.plot.PlotMapView(model = gwf, layer = 0)`
     `14 #qm = pmv.plot_array(leak)#, vmin = 0, vmax = 10) `
     `15 #pmv.contour_array(head, levels = levels)`

`File /Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/flopy/plot/map.py:103, in` `PlotMapView.plot_array(self, a, masked_values, **kwargs)`
    `100     a = np.array(a)`
    `102 # Use the model grid to pass back an array of the correct shape`
`--> 103 plotarray = self.mg.get_plottable_layer_array(a, self.layer)`
    `105 # if masked_values are provided mask the plotting array`
    `106 if masked_values is not None:`

`File /Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/flopy/discretization/structuredgrid.py:1562, in StructuredGrid.get_plottable_layer_array(self, a, layer)`
   `1560         plotarray = plotarray[layer, :, :]`
   `1561 else:`
`-> 1562     raise Exception("Array to plot must be of dimension 1, 2, or 3")`
   `1563 msg = f"{plotarray.shape} /= {required_shape}"`
   `1564 assert plotarray.shape == required_shape, msg`
`Exception: Array to plot must be of dimension 1, 2, or 3`

Then I checked the `leak` array. It prints like:
`[masked_array(`
  `data=[[[--, --, --, --, 0.26582127809524536, --, --, --, --, --],`
        ` [--, --, --, --, 0.2658206522464752, 0.3200981914997101, --,`
          `--, --, --],`
         `[--, --, --, --, --, 0.32008233666419983, --, --, --, --],`
         `[--, --, --, --, --, 0.320082426071167, 0.38010478019714355,`
          `--, --, --],`
         `[--, --, --, --, --, --, 0.3800913393497467, --, --, --],`
         `[--, --, --, --, --, --, 0.38009142875671387, --, --, --],`
         `[--, --, --, --, --, --, 0.380103200674057, 0.4000000059604645,`
          `--, --],`
         `[--, --, --, --, --, --, --, 0.4000000059604645, --, --],`
         `[--, --, --, --, --, --, --, 0.4000000059604645, --, --],`
         `[--, --, --, --, --, --, --, 0.4000000059604645, --, --]]],`
  `mask=[[[ True,  True,  True,  True, False,  True,  True,  True,  True,`
           `True],`
         `[ True,  True,  True,  True, False, False,  True,  True,  True,`
          ` True],`
         `[ True,  True,  True,  True,  True, False,  True,  True,  True,`
           `True],`
         `[ True,  True,  True,  True,  True, False, False,  True,  True,`
          ` True],`
        `[ True,  True,  True,  True,  True,  True, False,  True,  True,`
          ` True],`
         `[ True,  True,  True,  True,  True,  True, False,  True,  True,`
           `True],`
         `[ True,  True,  True,  True,  True,  True, False, False,  True,`
           `True],`
         `[ True,  True,  True,  True,  True,  True,  True, False,  True,`
           `True],`
         `[ True,  True,  True,  True,  True,  True,  True, False,  True,`
           `True],`
         `[ True,  True,  True,  True,  True,  True,  True, False,  True,`
           `True]]],`
  `fill_value=1e+20,`
  `dtype=float32)]`

How can I fix it?

**Reply by jlarsen-usgs** [reply]
The array looks like it's being returned within a list (ie. 4D, with time as the outer dimension). If you remove the time dimension as such you should be able to plot this:

```python
leak = leak[0]
pc = pmv.plot_array(leak, masked_values=[0,])
```

**Reply by QingHe-UT** [reply]
Thank you so much. It works! 
One more question is that, is the river leakage I extract here the drainage from the river to the aquifer (or the recharge from the aquifer to river)? If so, how can I know in which way the water flows?

I see in some literature the analytical flow pattern is often used to represent river-aquifer interactions (see the figure below for example). Using flopy, how can I plot such an analytical figure?
<img width="750" alt="Screen Shot 2023-02-04 at 19 14 08" src="https://user-images.githubusercontent.com/115207091/216762310-d5d0e5d0-a6d8-494b-bb68-b681631d5578.png">
source: https://journalpsij.com/index.php/PSIJ/article/view/586/1171
